File: smoke_tests/dart/insecure_network.dart
// Insecure Network Configuration vulnerabilities in Dart

import 'dart:io';
import 'package:http/http.dart' as http;

// Test 1: HTTP instead of HTTPS
Future<String> vulnerableHttp(String endpoint) async {
  // VULNERABLE: Using HTTP instead of HTTPS
  var response = await http.get(Uri.parse('http://api.example.com/$endpoint'));
  return response.body;
}

// Test 2: Disable certificate verification
Future<void> vulnerableNoCertVerify() async {
  // VULNERABLE: Disabling SSL certificate verification
  var client = HttpClient();
  client.badCertificateCallback = (cert, host, port) => true;
}

// Test 3: Accept any certificate
class VulnerableCertificateCheck {
  bool onBadCertificate(X509Certificate cert, String host, int port) {
    // VULNERABLE: Accepting all certificates
    return true;
  }
}

// Test 4: Weak TLS version
Future<void> vulnerableWeakTls() async {
  // VULNERABLE: Not enforcing TLS 1.2+
  var context = SecurityContext();
  // Not setting minimum TLS version
}

// Test 5: No certificate pinning
Future<String> vulnerableNoPinning(String url) async {
  // VULNERABLE: No certificate pinning
  var response = await http.get(Uri.parse(url));
  return response.body;
}

// Test 6: Sensitive data over HTTP
Future<void> vulnerableSensitiveHttp(String password) async {
  // VULNERABLE: Sending password over HTTP
  await http.post(
    Uri.parse('http://api.example.com/login'),
    body: {'password': password},
  );
}

// Test 7: Basic auth over HTTP
Future<void> vulnerableBasicAuthHttp(String username, String password) async {
  // VULNERABLE: Basic auth over HTTP
  var client = HttpClient();
  var request = await client.getUrl(Uri.parse('http://api.example.com/data'));
  request.headers.add('Authorization', 'Basic $username:$password');
}

// Test 8: API key in HTTP request
Future<void> vulnerableApiKeyHttp(String apiKey) async {
  // VULNERABLE: API key over HTTP
  await http.get(
    Uri.parse('http://api.example.com/data'),
    headers: {'X-API-Key': apiKey},
  );
}

// Test 9: Unencrypted WebSocket
Future<WebSocket> vulnerableWs(String host) async {
  // VULNERABLE: Unencrypted WebSocket
  return await WebSocket.connect('ws://$host/socket');
}

// Test 10: Ignoring SSL errors
Future<String> vulnerableIgnoreSsl(String url) async {
  // VULNERABLE: Ignoring all SSL errors
  var client = HttpClient();
  client.badCertificateCallback = (_, __, ___) => true;
  var request = await client.getUrl(Uri.parse(url));
  var response = await request.close();
  return await response.transform(SystemEncoding().decoder).join();
}

// Test 11: No hostname verification
Future<void> vulnerableNoHostnameVerify() async {
  // VULNERABLE: Would not verify hostname matches certificate
  var client = HttpClient();
  // Not verifying hostname
}

// Test 12: Hardcoded proxy
Future<void> vulnerableHardcodedProxy() async {
  // VULNERABLE: Hardcoded proxy (could be malicious)
  var client = HttpClient();
  client.findProxy = (uri) => 'PROXY proxy.example.com:8080';
}

// Test 13: Clear text traffic allowed
class NetworkConfig {
  // VULNERABLE: Allowing clear text traffic
  static const bool allowClearText = true;
}

// Test 14: Socket without TLS
Future<Socket> vulnerableRawSocket(String host, int port) async {
  // VULNERABLE: Raw socket without TLS
  return await Socket.connect(host, port);
}

// Test 15: FTP connection
Future<void> vulnerableFtp(String host, String username, String password) async {
  // VULNERABLE: FTP is unencrypted
  var socket = await Socket.connect(host, 21);
  socket.write('USER $username\r\n');
  socket.write('PASS $password\r\n');
}

File: smoke_tests/swift/insecure_network.swift
// Insecure Network Communication vulnerabilities in Swift
import Foundation

class InsecureNetwork {

    // Test 1: HTTP instead of HTTPS
    func fetchOverHttp() {
        // VULNERABLE: Using HTTP
        let url = URL(string: "http://api.example.com/data")!
        let task = URLSession.shared.dataTask(with: url) { data, response, error in
            print("Received data")
        }
        task.resume()
    }

    // Test 2: Disabling ATS (App Transport Security)
    // In Info.plist:
    // <key>NSAppTransportSecurity</key>
    // <dict>
    //   <key>NSAllowsArbitraryLoads</key>
    //   <true/>  <!-- VULNERABLE -->
    // </dict>

    // Test 3: Trust all certificates
    class TrustAllDelegate: NSObject, URLSessionDelegate {
        func urlSession(_ session: URLSession,
                       didReceive challenge: URLAuthenticationChallenge,
                       completionHandler: @escaping (URLSession.AuthChallengeDisposition, URLCredential?) -> Void) {
            // VULNERABLE: Accepting all certificates
            if let trust = challenge.protectionSpace.serverTrust {
                completionHandler(.useCredential, URLCredential(trust: trust))
            }
        }
    }

    // Test 4: Ignoring certificate errors
    func fetchIgnoringCerts(url: URL) {
        let config = URLSessionConfiguration.default
        let delegate = TrustAllDelegate()
        // VULNERABLE: Using delegate that trusts all certs
        let session = URLSession(configuration: config, delegate: delegate, delegateQueue: nil)
        session.dataTask(with: url).resume()
    }

    // Test 5: No certificate pinning
    func fetchWithoutPinning(url: URL) {
        // VULNERABLE: No certificate pinning implemented
        URLSession.shared.dataTask(with: url) { data, response, error in
            // Process response
        }.resume()
    }

    // Test 6: Credentials in URL
    func fetchWithCredsInUrl() {
        // VULNERABLE: Credentials visible in URL
        let url = URL(string: "https://user:password@api.example.com/data")!
        URLSession.shared.dataTask(with: url).resume()
    }

    // Test 7: Sensitive data over unencrypted WebSocket
    func connectWebSocket() {
        // VULNERABLE: Using ws:// instead of wss://
        let url = URL(string: "ws://example.com/socket")!
        let task = URLSession.shared.webSocketTask(with: url)
        task.resume()
    }

    // Test 8: Man-in-the-middle vulnerable
    func connectToBackend() {
        // VULNERABLE: No additional security measures
        var request = URLRequest(url: URL(string: "https://api.example.com")!)
        // Should implement certificate pinning
        URLSession.shared.dataTask(with: request).resume()
    }

    // Test 9: Cleartext traffic exception
    // In Info.plist:
    // <key>NSAppTransportSecurity</key>
    // <dict>
    //   <key>NSExceptionDomains</key>
    //   <dict>
    //     <key>example.com</key>
    //     <dict>
    //       <key>NSExceptionAllowsInsecureHTTPLoads</key>
    //       <true/>  <!-- VULNERABLE -->
    //     </dict>
    //   </dict>
    // </dict>

    // Test 10: Weak TLS version
    func configureWeakTLS() -> URLSessionConfiguration {
        let config = URLSessionConfiguration.default
        // VULNERABLE: Would allow weak TLS if possible
        // Modern iOS doesn't allow this but older versions might
        return config
    }

    // Test 11: API key in request header
    func fetchWithApiKey() {
        var request = URLRequest(url: URL(string: "https://api.example.com")!)
        // VULNERABLE: API key could be intercepted without proper security
        request.setValue("sk-secret-api-key-12345", forHTTPHeaderField: "X-API-Key")
        URLSession.shared.dataTask(with: request).resume()
    }

    // Test 12: Logging network responses
    func fetchAndLog(url: URL) {
        URLSession.shared.dataTask(with: url) { data, response, error in
            // VULNERABLE: Logging potentially sensitive response
            if let data = data, let str = String(data: data, encoding: .utf8) {
                print("Response: \(str)")
                NSLog("Response data: %@", str)
            }
        }.resume()
    }
}
